### ISSUE TITLE
Licence not removed on course completion - no clear/delete option - cannot retake course

### ISSUE THREAD

**Issue by jlb224** [iomad/iomad#1501]
We have 5 users who have successfully completed a course but the licence has not been removed. They are no longer visible in the participants list. 
There is no Clear or Delete button next to their name so we cannot remove the licence, nor unenrol them from the course (although they are not enrolled anyway). 
This means the user is unable to retake the certificate. This is an urgent issue for us. 

It looks like all users were enrolled and completed course on same day (not same day as each other) and course was set to auto-unerol 1 day after completion. 

In local_iomad_track I have removed the licence manually from licensename and licenseid (in a test database) and removed the row in local_report_usr_lic_allocs but the user name is still showing with asterisk (as if licence is still applied) so cannot be moved to list of potential users.

Please can you advise, or point in right direction to debug.

Version 3.8.5.

**Comment by turf212** [comment]
When you say they have successfully completed the course, have they completed the activities in the course only or have they met the course completion requirements for the course in question and were marked as complete for it within the Moodle course logs?

How were the user's enrolments removed from them for the course?  Was this through the license settings, through the the license enrolment method settings, by hand, or via the settings for the course in the Manage Iomad course settings page?

For the buttons to appear within the reports the user needs to be marked as complete within the local_iomad_track table.  This happens when the Moodle course_complete event is fired for that course/user combination which is generated when the course completion criteria is met within the course by that user.

The clear down of the user within the course to allow them to retake is either manual using the clear/delete buttons or is automatically driven by the Iomad course settings for that course.  (You need to set an expires after value which is the upper limit after which you can set the enrolment exipres which is the earliest value or the value when to send the "your training has expired" email which is the middle value.  It's explained in the ?s).  On the appropriate date being hit the user's interactions with the course are cleared down and their license is marked as completed with (companylicense_users timecompleted value is set for that license/course/user).  This then allows the user to be re-allocated a license for that course and undertake the course again from a clean state.

The clear/delete buttons also update the license record in companylicense_users too.

It sounds like you didn't have course completion criteria set up so that the course completion event was never fired/captured and Iomad doesn't believe that the course is completed.  The user was then removed from the course from the license enrolment.

If this is the case - TO FIX:

- Set up your course completion criteria in the course properly.
- For the affected users update the local_iomad_track table so that they have on appropriate timestamp value for the course.
- This will then prompt the clear button to show which you can use to clear the use from the course and mark their license as finished with, allowing you to provide them with another license to resit the course.

**Comment by jlb224** [comment]
Thanks for your reply. I am unsure of the history of the course / settings because I have only recently started working on the site but I looked in the local_iomad_track table and their completion and score was showing, but timecompleted and score is not present in mdl_companylicense_users. I have manually added the data for these two fields using the data present in local_iomad_track but I dont know 'why' this is happening, and further users (from another company / course) have been sent to me today with the same issue. Do you have any idea what would make that happen?

**Comment by turf212** [comment]
If you go to the course itself and go to course administration (or the cog depending on theme) and click on Course completion, do you have any criteria set up for the course to be marked as complete within Moodle.  If you have no criteria (which it sounds like is the case) then Moodle will not mark anyone as complete, causing the situation you are in.

**Comment by jlb224** [comment]
Yes course completion is set on the course. All activities (except end of course feedback) + course grade of 80 are set as course completion conditions. 

it looks like these users successfully complete the course. And are unenrolled (as in they no longer show in participants list). But iomad is not classing them as complete/unenrolled - they remain as licenced. the iomad tracking table sees the completion/score. The licence table does not. I dont know where else. I manually added the completion to the licence table which allowed them to be re-enrolled but apparently the course is showing all of their old responses and thinks they have completed it already. 

I have only just started working on the site but I do know that they recently upgraded to 3.8. And because the course was set to unenrol after 300 days (but they needed to renerol these people sooner, and the clear / delete buttons werent showing) they changed to unenrol after 1 day, but after 1 day it didnt happen. Hence the DB change but it hasnt resolved anything. And these exact situations (with no clear/delete button) keep occurring on different users/different courses.

**Comment by turf212** [comment]
So, to confirm:

Moodle marks the user as complete on the course
User report show user as complete for the course (data filled when Moodle completion is met)
User is unenrolled from the course (what controls this?  Is it from the license definition or have you changed the license authentication instance?)
Clear/delete buttons are not being seen.
companylicense_user record has timecompleted as NULL (expected as only happens when user is cleared from the course).

**Comment by jlb224** [comment]
Moodle marks the user as complete on the course **YES**
User report show user as complete for the course (data filled when Moodle completion is met) **YES**
User is unenrolled from the course (what controls this? Is it from the license definition or have you changed the license authentication instance?) **NO**. The users are not being unenrolled from the course after X amount of days, as we would expect. We control this function via Dashboard > Courses > Manage Iomad Course Settings. They are not being unenrolled and there is no Clear / Delete button for them (where I said above they were not present in Participants list, I have since discovered that was because the admin tried to unenrol manually when faced with no autoenrol, and no buttons - so them not being present there is unrelated I believe).
Clear/delete buttons are not being seen. **YES**
companylicense_user record has timecompleted as NULL (expected as only happens when user is cleared from the course). **YES**

About the auto unenrol, we think that has not worked since we upgraded to 3.8. It is usually set to unenrol after 300 days. This was changed to 1 day but the users have not been unenrolled. The licence is set to expire after 30 days, and end in 2021.

**Comment by turf212** [comment]
One last thing then, is the course completion event being fired for the user in the course at the time (or just after) they are meeting the course completion criteria?

The unenrolment happens after X days after Iomad has recorded their completion.  The buttons appear once Iomad has recorded their completion.  Iomad is not recording their completion and it's this event which it's listening on.  If that event is not being fired then it's a course set up issue.  If it is then it could be something else you have installed which is also listening for that event, fails, preventing the Iomad listener from triggering.

**Comment by jlb224** [comment]
I have setup on a brand new install (3.8.6) so it is easier to see what is going on (because whilst I know Moodle i am new to iomad) and have created one company, two users, a course with quiz with receive grade to complete as activity completion, allocated licence, logged in as user and taken quiz. I can see my 100% completion but no grade and no time completed in DB. What could be wrong in the completion settings? I have kept it as basic as can be. 

See below: course 2. (course 1 only used labels, no graded activity, and something weird happened with it. I left all iomad course settings as 0 (so it shouldn't unenrol, but it did unenrol, and no buttons to clear course). So I reset up using a graded activity which is in course 2.

![image](https://user-images.githubusercontent.com/26649166/101845853-9ac12c80-3b47-11eb-95aa-5600bcd080cf.png)

**Comment by jlb224** [comment]
The completion data does seem to be showing in timecompleted in local_iomad_track table. Is that where you mean by iomad recording their completion?

**Comment by turf212** [comment]
Yes that is what I mean.  There should be a timecompleted value which matches the timecompleted value in mdl_course_completions for the same userid and course/courseid.  If that is there then there should be no reason as to why the buttons are not showing.  Check with a different theme in case you have set some other buttons to display:none; and it's impacted these ones.

**Comment by jlb224** [comment]
I understand. I checked on fresh install of iomad 3.8.6 and can see that on first cron run after completion the completion status is added to user report which leaves it like course 3 with status 100% but no completion data, and after second run the completion date, valid until, and grade are populated like course 2. I guess an iomad scheduled task is running on cron run 2 (so is picking up the Moodle completion data added in cron run 1). We do have issues where it stays like course3. Could this be related to the buttons not appearing? As in, from what you said previously "If it is then it could be something else you have installed which is also listening for that event, fails, preventing the Iomad listener from triggering." and if so, how do I go about debugging that to find what is stopping it triggering? It seems like these might be related and there are indeed custom plugins on the site. I looked in iomad/local/iomad/classes/task/cron_task.php but it didnt look like it was doing anything in this area particularly. 

Are both the buttons appearing and the completion data being added to user report linked to the running of a particular scheduled task? 
![image](https://user-images.githubusercontent.com/26649166/102069974-645d0900-3df6-11eb-9efc-090ce3a73e07.png)
